### PR TITLE
Fixes #37227 - start the FreeIps service

### PR DESCRIPTION
**Pull request by nofaralfasi**

### CODE REVIEW

File: lib/smart_proxy_dhcp_infoblox/dhcp_infoblox_plugin.rb [hunk 1]
```diff
@@ -17,6 +17,8 @@ class Plugin < ::Proxy::Provider
     load_classes ::Proxy::DHCP::Infoblox::PluginConfiguration
     load_dependency_injection_wirings ::Proxy::DHCP::Infoblox::PluginConfiguration
 
+    start_services :unused_ips
```
Comment (timdeluxe) [inline lib/smart_proxy_dhcp_infoblox/dhcp_infoblox_plugin.rb:20]: @nofaralfasi Can you move this after the following validates? I think it makes no sense to start the service, when one of the validation fails. Thanks!

**Review by timdeluxe — CHANGES_REQUESTED** [review]
Thanks @ekohl for the detailled analysis in the bugzilla comments.

Do i get it right, that the issue is not that the "Suggest new" does not suggest new IPs at all, but if hit often it "eats up" free ips by allocating them (within foreman), but does not release them after some time, correct?

If yes, that is an interesting edge case. Never thought that someone might hit "Suggest new" that often to hit this bug. I could confirm the bug with a small subnet. However i need some time to deploy the fix in our test environment to confirm it fixes the bug.

**Comment by ekohl** [comment]
That is correct. It would be better if Foreman provided a way to say "after I submit the form, suggest an IP" instead of the current way. That would reduce the impact a lot. Essentially the Suggest new functionality allows users to DoS themselves or others

File: lib/smart_proxy_dhcp_infoblox/dhcp_infoblox_plugin.rb (hunk 1 above)
Comment (ekohl) [reply lib/smart_proxy_dhcp_infoblox/dhcp_infoblox_plugin.rb]: The order shouldn't really matter since all it does is store the services to be started: https://github.com/theforeman/smart-proxy/blob/e8fe59f83843ca619264e8d4757bc99fcde93fa9/lib/proxy/pluggable.rb#L88-L90

The services are only started once the plugin is fully wired up. If you want to know the details:
* The launcher initializes the plugins using the `PluginInitializer`: https://github.com/theforeman/smart-proxy/blob/e8fe59f83843ca619264e8d4757bc99fcde93fa9/lib/launcher.rb#L129
* The `PluginInitializer` configures all plugins: https://github.com/theforeman/smart-proxy/blob/e8fe59f83843ca619264e8d4757bc99fcde93fa9/lib/proxy/plugin_initializer.rb#L159-L197
* Eventually it configures each `PluginGroup` using the module loader: https://github.com/theforeman/smart-proxy/blob/e8fe59f83843ca619264e8d4757bc99fcde93fa9/lib/proxy/plugin_initializer.rb#L107-L114
* The `DefaultRuntimeConfigurationLoader` wires up the dependencies before starting services: https://github.com/theforeman/smart-proxy/blob/e8fe59f83843ca619264e8d4757bc99fcde93fa9/lib/proxy/plugin_initializer.rb#L238-L239

**Comment by timdeluxe** [comment]
@nofaralfasi I see your review-request, but i kindly request you to be a bit patient. This week it is absolutely not possible due to full of work/todos at work time and leisure time. It is still on my todo list and hopefully i can do it next week.
(just to remind: i want to verify it on a foreman instance with the updated plugin code, instead of just reviewing the code)

The [bugzilla issue](https://bugzilla.redhat.com/show_bug.cgi?id=2207698) also is open since a "few" days, so i hope the additional ones here do not matter, or do they?

**Comment by nofaralfasi** [comment]
@timdeluxe absolutely no rush. I simply wanted to confirm that you received my latest update :)

**Review by timdeluxe — APPROVED** [review]
Sorry for the big delay. Today i got the chance to test it.
I can confirm that with the change, the cleanup will be done and without it won't. Only with the change i get the expected
`2024-03-19T14:34:06  [D] Starting allocated ip addresses cleanup pass...`
in the log (log level debug!) and allocated addresses are indeed cleaned up.

The cleanup however only works in a timely manner, when one sets the blacklist interval rather small, @ekohl has mentioned this already in the bugzilla issue. Quote:

> What you can see is that it looks for an available IP by taking the set of IPs it's allowed to allocate within the subnet, subtract the already allocated. The IP it found is temporarily marked as allocated. This period is defined as the blacklist interval, defaulting to 30 * 60 (so 30 minutes). Cleanup of the reservations pool happens every minute. The is no mechanism to manually mark a previously suggested IP address as free again.
>
>You can actually override this blacklist duration, though there is no installer option for it. This would be done in by adding :blacklist_duration_minutes to /etc/foreman-proxy/settings.d/dhcp_isc.yml

Without this blacklist duration change, the cleanup is done, but it does not cleanup any ips, if the blacklist time (default 30 minutes) has not passed. @nofaralfasi I think that needs to be extra highlighted to the reporter over at bugzilla!

@nofaralfasi I'll approve here now and kickstart the release process, but please be patient - it can take a while (the release is not only in my hands)!

**Comment by timdeluxe** [comment]
gem is released: https://rubygems.org/gems/smart_proxy_dhcp_infoblox
for release as packages we need to wait for weekly update packages automation: https://github.com/theforeman/foreman-packaging/actions after which i can do the required cherry picking. I've set a reminder for myself.

**Comment by ekohl** [comment]
Thanks for also doing the release. I've kicked off the weekly scheduled job now so https://github.com/theforeman/foreman-packaging/pull/10602 and https://github.com/theforeman/foreman-packaging/pull/10603 are the PRs to watch.

**Comment by timdeluxe** [comment]
Thanks.
The release worked by itself thanks to you adding the github workflow recently 😉
I just committed, tagged and pushed...
Will watch the PRs.

**Comment by timdeluxe** [comment]
@nofaralfasi 
Packaging automation did its thing, but that didn't include the new release of the plugin to the current foreman releases.

For that i had to cherry-pick (which gives me headaches, but i finally found my way). These are the PRs:
https://github.com/theforeman/foreman-packaging/pull/10657
https://github.com/theforeman/foreman-packaging/pull/10656
https://github.com/theforeman/foreman-packaging/pull/10655
https://github.com/theforeman/foreman-packaging/pull/10654
https://github.com/theforeman/foreman-packaging/pull/10653
https://github.com/theforeman/foreman-packaging/pull/10652

This should package it for Foreman 3.8, 3.9 and 3.10. Didn't do it for older releases as i consider them as not supported anymore. Even for 3.8 i doubt it.
If i am not wrong, the packages should be automatically published for the relevant foreman release shortly after those PRs are merged. It shouldn't require another foreman release.

I think you know the best, when and how to update bugzilla or redmine bug entries...

**Comment by nofaralfasi** [comment]
Thank you @timdeluxe! I'll take care of updating everything else.

**Comment by ekohl** [comment]
> For that i had to cherry-pick (which gives me headaches, but i finally found my way).

Yes, it's always a bit tedious. I recently wrote a small script to simplify things:
```bash
#!/bin/bash

COMMITS=$1 ; shift
BRANCHES=$*
REMOTE=origin

set -e

if [[ -z $BRANCHES ]] ; then
        echo "Usage: $0 COMMITS BRANCH [BRANCH]"
        exit 1
fi

BRANCH_SUFFIX=$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 13 ; echo '')

git fetch "$REMOTE"

for BRANCH in $BRANCHES ; do
        git checkout -b "$BRANCH-pick-$BRANCH_SUFFIX" "${REMOTE}/${BRANCH}"
        git cherry-pick -x "$COMMITS"
        git phr -b "${BRANCH}" --no-edit
done
```

It currently only works for a single argument, but you can use it as `foreman-packaging-pick some_commit_sha rpm/3.{9..10}` (relying on bash to expand the range for you).

It uses `git phr` which is my alias:
```sh
$ git phr --help
'phr' is aliased to '!git ph && git pr'
$ git ph --help
'ph' is aliased to '!git push $(whoami) HEAD --set-upstream --quiet'
$ git pr --help
'pr' is aliased to '!hub pull-request'
```

> This should package it for Foreman 3.8, 3.9 and 3.10. Didn't do it for older releases as i consider them as not supported anymore. Even for 3.8 i doubt it.

You're right, 3.8 is EOL now that 3.10 is GA.

**Comment by timdeluxe** [comment]
@ekohl Thanks for the helpfull snippets, might use them next time.

@nofaralfasi Looks like it is now complete, at least on my system apt now wants to update the plugin:
`   ruby-smart-proxy-dhcp-infoblox (0.0.17-2 => 0.0.18-1~fm3.9)`
